JavaScript/JavaScript-Classes.md

```diff
@@ -0,0 +1,171 @@
+
+If you want to create multiple objects with similar methods and properties in JavaScript, the best options are functional classes, or the constructor function. There are four styles of functional classes.
+
+ * Functional
+
+ * Functional-Shared
+
+ * Prototypal
+
+ * Pseudoclassical
+
+Each of these four is a different approach to object construction; they all accomplish the same thing in a slightly different way.
+
+Take for example, a street, which has many houses. Every house on the street requires a colour for the house, a front door with methods to open and close the door. A very simple example of where houses may be thought of as objects. 
+
+We can implement functional classes here, as every instance created by the function will have the attributes of house colour and the methods to open and close the door.  
+
+From here we can potentially create hundreds, thousands or even more new house objects, which are similar.  So each functional class performs the same basic steps to generate the new houses.
+
+* generate an object
+* assign properties
+* add methods
+* return that object
+
+Each of the four styles has its pros and cons.
+
+####Functional
+Firstly **functional style** is **fast, very simple and clear in object creation**.  However all methods will be duplicated for every single object created, so a new function is stored in memory every time the object is generated.
+
+````
+var House = function(colour){
+	var obj = {}; // generate object
+	
+	obj.colour = colour;
+	obj.door = 'open'; //assign properties
+	
+	obj.openDoor = function(){
+		obj.door = 'open';
+	}; //explicitly define or borrow methods
+	
+	obj.closeDoor = function(){
+		obj.door = 'closed;
+	}; //explicitly define or borrow methods
+	
+	return obj;
+}
+
+var house = House('red'); //instantiation pattern
+	
+````
+
+####Functional-Shared
+Secondly **functional-shared**, which is better for memory management as it **utilises a single repository for the methods**. Every time an object is created, pointers are generated, so every new object will point back to the houseMethods object, where the functions are stored one single time in memory.  
+
+````
+var House = function(colour){
+	var obj = {}; // generate object
+	
+	obj.colour = colour;
+	obj.door = 'open'; //assign properties
+	
+	obj.open = houseMethods.open;
+	obj.close = houseMethods.close; //explicitly define or borrow methods
+	
+	return obj;
+};
+
+var houseMethods = {};
+
+houseMethods.openDoor = function(){
+	this.door = 'open'; //Add method to delegate fallback object
+};
+
+houseMethods.closeDoor = function(){
+	this.door = 'closed'; //Add method to delegate fallback object
+
+};
+
+var house = House('red'); //instantiation pattern
+
+````
+This is not as efficient as delegating through fallback.
```

Comment:
I am not quite clear here what you mean by 'delegating'. Does this mean retrieving properties from the prototype?